python-checks/src/main/java/org/sonar/python/checks/TfFunctionRecursivityCheck.java

```diff
@@ -0,0 +1,114 @@
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.symbols.Usage;
+import org.sonar.plugins.python.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Decorator;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6908")
+public class TfFunctionRecursivityCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Remove this recursive call.";
+  private static final String SECONDARY_MESSAGE = "Recursive call is here.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FUNCDEF, TfFunctionRecursivityCheck::checkFunctionDef);
+  }
+
+  private static void checkFunctionDef(SubscriptionContext context) {
+    FunctionDef functionDef = (FunctionDef) context.syntaxNode();
+    if (!isTfFunction(functionDef)) {
+      return;
+    }
+    var selfSymbol = TreeUtils.getFunctionSymbolFromDef(functionDef);
+    if (selfSymbol == null) {
+      return;
+    }
+    var collector = new CallCollector(selfSymbol);
+    context.syntaxNode().accept(collector);
+    if (collector.expressionList.isEmpty()) {
+      return;
+    }
+    var issue = context.addIssue(functionDef.name(), MESSAGE);
+    collector.expressionList.forEach(call -> issue.secondary(call, SECONDARY_MESSAGE));
+  }
+
+  private static boolean isTfFunction(FunctionDef functionDefinition) {
+    return functionDefinition.decorators().stream()
+      .map(Decorator::expression).map(TreeUtils::getSymbolFromTree)
+      .filter(Optional::isPresent).map(Optional::get)
+      .map(Symbol::fullyQualifiedName)
+      .filter(Objects::nonNull)
+      .anyMatch("tensorflow.function"::equals);
+  }
+
+  private static class CallCollector extends BaseTreeVisitor {
+    private final Symbol originalSymbol;
+    List<Expression> expressionList = new ArrayList<>();
+
+    Set<FunctionDef> visited = new HashSet<>();
+
+    private CallCollector(@Nonnull Symbol originalSymbol) {
+      this.originalSymbol = originalSymbol;
+    }
+
+    @Override
+    public void visitCallExpression(CallExpression callExpression) {
+      Symbol symbol = TreeUtils.getSymbolFromTree(callExpression.callee()).orElse(null);
+      if (symbol == null) {
+        return;
+      }
+      if (symbol.equals(originalSymbol)) {
+        expressionList.add(callExpression.callee());
+      }
+      if (symbol.is(Symbol.Kind.FUNCTION)) {
```

Comment:
I believe having a `else if` here fixes the `StackOverflowError` you've been having in the current tests.

Some can still be trigger if two functions that do not match the original symbol call each other. Still, I assume it is not the intention to visit the same function declaration again.